source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,927 @@
+                                   fe_data.enrichment[base_no].size()));
+      for (unsigned int m=0; m < base_no_mult_local_enriched_dofs[base_no].size(); m++)
+        {
+          if (flags & update_hessians)
+            {
+              Assert (fe_data.enrichment[base_no][m].hessians.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].hessians.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].hessians[q]  = enrichments[base_no-1][m](cell)->hessian (mapping_data.quadrature_points[q]);
+            }
+
+          if (flags & update_gradients)
+            {
+              Assert (fe_data.enrichment[base_no][m].gradients.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].gradients.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].gradients[q] = enrichments[base_no-1][m](cell)->gradient(mapping_data.quadrature_points[q]);
+            }
+
+          if (flags & update_values)
+            {
+              Assert (fe_data.enrichment[base_no][m].values.size() == n_q_points,
+                      ExcDimensionMismatch(fe_data.enrichment[base_no][m].values.size(),
+                                           n_q_points));
+              for (unsigned int q=0; q<n_q_points; q++)
+                fe_data.enrichment[base_no][m].values[q]    = enrichments[base_no-1][m](cell)->value   (mapping_data.quadrature_points[q]);
+            }
+        }
+    }
+
+  // Finally, update the standard data stored in output_data
+  // by expanding the product rule for enrichment function.
+  // note that the order if important, namely
+  // output_data.shape_XYZ contains values of standard FEM and we overwrite
+  // it with the updated one in the following order: hessians -> gradients -> values
+  if (flags & update_hessians)
+    {
+      Tensor<2, spacedim> grad_grad, grad_grad_r;
+      Tensor<2,spacedim,double> sym_grad_grad, sym_grad_grad_r;
+      Tensor<1,spacedim> grad_r;
+      for (unsigned int base_no = 1; base_no < this->n_base_elements(); base_no++)
+        {
+          for (unsigned int m=0; m < base_no_mult_local_enriched_dofs[base_no].size(); m++)
+            for (unsigned int i=0; i < base_no_mult_local_enriched_dofs[base_no][m].size(); i++)
+              {
+                const unsigned int &enriched_dof = base_no_mult_local_enriched_dofs[base_no][m][i];
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    grad_grad =  outer_product(output_data.shape_gradients[enriched_dof][q],  fe_data.enrichment[base_no][m].gradients[q]);
+                    sym_grad_grad = symmetrize (grad_grad) * 2.0; // symmetrize does [s+s^T]/2
```

Comment:
i would say it is not, but last time i did derivations that was the outcome of the chain rule